src/jdk.crypto.cryptoki/unix/native/libj2pkcs11/p11_md.c

```diff
@@ -264,18 +264,19 @@ JNIEXPORT jobject JNICALL Java_sun_security_pkcs11_wrapper_PKCS11_connect
  * Method:    disconnect
  * Signature: ()V
  */
-JNIEXPORT void JNICALL Java_sun_security_pkcs11_wrapper_PKCS11_disconnect
-    (JNIEnv *env, jobject obj)
-{
-    ModuleData *moduleData;
+JNIEXPORT void JNICALL Java_sun_security_pkcs11_wrapper_PKCS11_disconnect(
+        JNIEnv *env, jclass thisClass, jlong ckpNativeData) {
+
     TRACE0("DEBUG: disconnecting module...");
-    moduleData = removeModuleEntry(env, obj);
+    if (ckpNativeData != 0L) {
+        ModuleData *moduleData = jlong_to_ptr(ckpNativeData);
 
-    if (moduleData != NULL) {
-        dlclose(moduleData->hModule);
+        if (moduleData != NULL) {
```

Comment:
The check should be (moduleData->hModule != NULL)?

Reply:
Or should both be checked?

Reply:
I think it is safer to add the check for 'hModule'.

```
-        if (moduleData != NULL) {
+        if (moduleData != NULL && moduleData->hModule != NULL) {

```

Reply:
That is very safe -- we already checked that `ckpNativeData != 0L`

Reply:
Hm, I see now.